text/3498-lifetime-capture-rules-2024.md

```diff
@@ -0,0 +1,642 @@
+
+```rust
+#![feature(impl_trait_in_assoc_type)]
+
+trait Trait<'t> {
+    type Gat<'g> where 'g: 't; // Bound required by existing GAT rules.
+    fn foo<'f>(self, x: &'t (), y: &'f ()) -> Self::Gat<'f>;
+}
+
+struct Foo<'s>(&'s ());
+impl<'t, 's> Trait<'t> for Foo<'s> {
+    type Gat<'g> = impl Sized where 'g: 't;
+    //             ^^^^^^^^^^
+    //             ^ Captures:
+    //
+    //                 - `'g` from the GAT.
+    //                 - `'f` from the method signature (via the GAT).
+    //                 - `'t` from the outer impl and a trait input.
+    //                 - `'s` from the outer impl and Self type.
+    fn foo<'f>(self, x: &'t (), y: &'f ()) -> Self::Gat<'f> {
+        (self, x, y)
+    }
+}
+```
+
+This updates and supersedes the behavior specified in [RFC 2071] and [RFC 2515].
+
+## Return position `impl Trait` in Trait (RPITIT)
+
+Under this RFC, when an associated function or method in a trait definition contains in its return type a return position `impl Trait` in trait (RPITIT), the impl of that item may capture in the returned opaque type, in all editions, all trait input type and lifetime parameters, all type and lifetime parameters present in the `Self` type, and all type and lifetime parameters in the associated function or method signature.
+
+When such an associated function or method in a trait definition provides a default implementation, the opaque return type will automatically capture all trait input type and lifetime parameters, all type and lifetime parameters present in the `Self` type, and all type and lifetime parameters in the associated function or method signature.
+
+In trait impls, return position `impl Trait` (RPIT), in all editions, will automatically capture all type and lifetime parameters from the outer impl and from the associated function or method signature.  This ensures that signatures are copyable from trait definitions to impls.
+
+For example:
+
+```rust
+#![feature(return_position_impl_trait_in_trait)]
+
+trait Trait<'t> {
+    fn foo<'f>(self, x: &'t (), y: &'f ()) -> impl Sized;
+    //                                        ^^^^^^^^^^
+    // Method signature lifetimes, trait input lifetimes, and
+    // lifetimes in the Self type may all be captured in this opaque
+    // type in the impl.
+}
+
+struct Foo<'s>(&'s ());
+impl<'t, 's> Trait<'t> for Foo<'s> {
+    fn foo<'f>(self, x: &'t (), y: &'f ()) -> impl Sized {
+        //                                    ^^^^^^^^^^
+        // The opaque type captures:
+        //
+        //   - `'f` from the method signature.
+        //   - `'t` from the outer impl and a trait input lifetime.
+        //   - `'s` from the outer impl and the Self type.
+        (self, x, y)
+    }
+}
+```
+
+This updates and supersedes the behavior specified in [RFC 3425].
+
+[RFC 3425]: https://github.com/rust-lang/rfcs/blob/master/text/3425-return-position-impl-trait-in-traits.md
+
+## `async fn` in trait (AFIT)
```

Comment:
I don't think there is any change to this feature; this behavior was documented e.g. [in this section](https://github.com/rust-lang/rfcs/blob/master/text/3425-return-position-impl-trait-in-traits.md#difference-in-scoping-rules-from-async-fn) of the RPITIT RFC.

Reply:
Is that true?  That section of the RFC presents this example:

```rust
trait Trait {
    async fn async_fn(&self);
}

impl Trait for MyType {
    fn async_fn(&self) -> impl Future<Output = ()> + '_ { .. }
}
```

Then it goes on to present how the `Captures` trick must be used.

Reply:
I'll mark this as resolved, but please unresolve it if there is something that should be changed here.